src/gwt/panmirror/src/editor/src/optional/ace/render_queue.ts

```diff
@@ -0,0 +1,215 @@
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+import { CodeBlockNodeView } from './ace';
+
+/**
+ * Represents a queue of Ace editor instances that are rendered asynchronously.
+ */
+export class AceRenderQueue {
+  private readonly renderQueue: CodeBlockNodeView[] = [];
+
+  private renderCompleted: boolean = false;
+  private renderTimer: number = 0;
+  private container?: HTMLElement;
+  private needsSort: boolean = true;
+  private bypass: number = 5;
+  private observer?: IntersectionObserver;
+  private visible: boolean = true;
+
+  /**
+   * Sets (or replaces) the scroll container hosting the render queue. The
+   * scroll container is used to prioritize the queue (i.e. objects in the
+   * viewport or close to it are to be given more priority).
+   * 
+   * @param container The HTML element of the scroll container
+   */
+  public setContainer(container: HTMLElement) {
+
+    // Handler for scroll events in the container
+    const handleScroll = (evt: Event) => {
+      // Whenever a scroll event occurs, we need to re-sort the queue so that
+      // objects in or closest to the viewport are given priority.
+      this.needsSort = true;
+
+      // If we don't think we're visible but we're scrolling and have height,
+      // then we are in fact visible. (This catches a case where the
+      // intsersection observer created below doesn't fire for visiblity
+      // changes.)
+      if (!this.visible && this.container && this.container.offsetHeight > 0) {
+        this.visible = true;
+        this.processRenderQueue();
+      }
+    };
+
+    // Skip if we're already looking at this container
+    if (this.container === container) {
+      return;
+    }
+
+    // Clean up handlers on any previous container
+    if (this.container) {
+      this.container.removeEventListener("scroll", handleScroll);
+    }
+    if (this.observer) {
+      this.observer.disconnect();
+    }
+
+    this.container = container;
+
+    // Create intersection observer to ensure that we don't needlessly churn
+    // through renders while offscreen.
+    this.observer = new IntersectionObserver((entries: IntersectionObserverEntry[]) => {
+      for (const entry of entries) {
+        if (entry.isIntersecting && !this.visible) {
+          // We just became visible; process the render queue now.
+          this.visible = true;
+          this.processRenderQueue();
+        }
+        if (!entry.isIntersecting && this.visible) {
+          // We just lost visibility; end processing of the render queue. (Note
+          // that we only do this when connected to the DOM as another reason
+          // the element can become invisible is ProseMirror swapping it out
+          // internally.)
+          if (this.container?.parentElement) {
+            this.visible = false;
+            if (this.renderTimer > 0) {
+              window.clearTimeout(this.renderTimer);
```

Comment:
Should `this.renderTimer` be reset to `0` here?

Reply:
It should, I've made that change.